File: apps/android/app/src/main/java/com/variance/nearby/screens/activity/ActivityService.kt
package com.variance.nearby.screens.activity

import com.variance.nearby.core.AppConstants
import com.variance.nearby.leansui.api.GraphQLSuiProvider
import com.variance.nearby.leansui.api.SuiNetwork
import kotlinx.coroutines.future.await
import org.swift.swiftkit.core.SwiftArena
import java.util.Optional

/**
 * IO boundary for the on-chain activity feed. Forwards to the shared `getActivity` (fetch + fold +
 * format), scoped to the configured balance coin, and maps the bridged result into plain models.
 */
class ActivityService(
    network: SuiNetwork,
    private val swiftArena: SwiftArena,
    private val coinType: String = AppConstants.USD_SUI_COIN_TYPE,
) {
    private val provider: GraphQLSuiProvider = GraphQLSuiProvider.init(network, swiftArena)

    /** One page of the address's activity, newest first. Pass `cursor` = null for the first page. */
    suspend fun activity(address: String, cursor: String?): ActivityPage {
        val feed = provider
            .getActivity(address, coinType, Optional.ofNullable(cursor), DEFAULT_LIMIT, swiftArena)
            .await()
        return feed.toPage(swiftArena)
    }

    private companion object {
        // Page size is bounded by the GraphQL query cost; the package default is 25.
        const val DEFAULT_LIMIT = 25L
    }
}

File: apps/ios/nearby/Features/Activity/ActivityService.swift
import Foundation
import LeanSuiApi

/// IO boundary for the on-chain activity feed. Wraps a GraphQL provider and forwards to the shared
/// `getActivity` (which does the fetch + fold + format), scoped to the configured balance coin.
actor ActivityService {
  private let provider: GraphQLSuiProvider
  private let coinType: String

  init(network: SuiNetworkKind, coinType: String) {
    self.provider = GraphQLSuiProvider(network: SuiNetwork(kind: network))
    self.coinType = coinType
  }

  /// One page of the address's activity, newest first. Pass `cursor` = nil for the first page.
  /// Page size is owned by the package (`getActivity`) since it's bounded by the GraphQL query cost.
  func activity(address: String, cursor: String?) async throws -> SuiActivityFeed {
    try await provider.getActivity(address: address, coinType: coinType, cursor: cursor)
  }
}
